### PR TITLE
feat(pipeline): The pipeline component manages action definitions 

### PR DESCRIPTION
**Pull request by kakj-go**
#### What this PR does / why we need it:
 The pipeline component manages action definitions 

Temporarily convert source to locations for actions query. Wait until all pipeline clients are used
pipeline definition, the definition of action is not queried from dicehub but from the action definition table of pipeline itself

#### Which issue(s) this PR fixes:

- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMTc0LDExNjRdLCJhc3NpZ25lZSI6WyIxMDAwNTYwIl19&id=301565&iterationID=1174&pId=0&type=REQUIREMENT)

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |        The pipeline component manages action definitions       |
| 🇨🇳 中文    |       pipeline 组件管理 action 定义       |

#### test
pipeline yaml:
```yaml
version: "1.1"
stages:
  - stage:
      - git-checkout:
          alias: git-checkout
          description: 代码仓库克隆
          version: "1.0"
          params:
            branch: ((gittar.branch))
            depth: 1
            password: ((gittar.password))
            uri: ((gittar.repo))
            username: ((gittar.username))
          timeout: 3600
  - stage:
      - java-builds:
          alias: java-build
          version: "1.0"
          params:
            build_cmd:
              - echo 1
            jdk_version: "8"
            workdir: .
  - stage:
      - java-build:
          alias: java-build1
          description: 针对 java 工程的编译打包任务
          version: "1.0"
          params:
            build_cmd:
              - echo 1
            jdk_version: "8"
            workdir: .
```

image

run result:

image

### CODE REVIEW

File: modules/pipeline/providers/action/action.service.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+func (s *actionService) CheckInternalClient(ctx context.Context) error {
+	if apis.GetInternalClient(ctx) == "" {
+		return fmt.Errorf("not internal client req")
+	}
+	return nil
+}
+
+func (s *actionService) List(ctx context.Context, req *pb.PipelineActionListRequest) (*pb.PipelineActionListResponse, error) {
+	if req.Locations == nil {
```
Comment (sfwn) [inline modules/pipeline/providers/action/action.service.go:47]: use `len()` is better for 0 length.

File: modules/pipeline/providers/action/action.service.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+func (s *actionService) Save(ctx context.Context, req *pb.PipelineActionSaveRequest) (*pb.PipelineActionSaveResponse, error) {
+	if err := s.CheckInternalClient(ctx); err != nil {
+		return nil, apierrors.ErrListPipelineAction.InternalError(err)
+	}
+
+	if req.Spec == "" {
+		return nil, apierrors.ErrSavePipelineAction.InvalidParameter("spec was empty")
+	}
+	if req.Dice == "" {
+		return nil, apierrors.ErrSavePipelineAction.InvalidParameter("dice was empty")
```
Comment (sfwn) [inline modules/pipeline/providers/action/action.service.go:86]: diceyml

File: modules/pipeline/providers/action/action.service.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+	}, nil
+}
+
+func (s *actionService) Save(ctx context.Context, req *pb.PipelineActionSaveRequest) (*pb.PipelineActionSaveResponse, error) {
+	if err := s.CheckInternalClient(ctx); err != nil {
+		return nil, apierrors.ErrListPipelineAction.InternalError(err)
+	}
+
+	if req.Spec == "" {
+		return nil, apierrors.ErrSavePipelineAction.InvalidParameter("spec was empty")
```
Comment (sfwn) [inline modules/pipeline/providers/action/action.service.go:83]: spec.yml

File: modules/pipeline/services/extmarketsvc/common.go [hunk 4]
```diff
@@ -25,3 +31,24 @@ func MakeActionTypeVersion(action *pipelineyml.Action) string {
...
 	return r
 }
+
+func MakeActionLocationsBySource(source apistructs.PipelineSource) []string {
+	var locations []string
+	switch source {
+	case apistructs.PipelineSourceCDPDev, apistructs.PipelineSourceCDPTest, apistructs.PipelineSourceCDPStaging, apistructs.PipelineSourceCDPProd, apistructs.PipelineSourceBigData:
+		locations = append(locations, apistructs.PipelineTypeFDP.String()+string(os.PathSeparator))
+	default:
+		locations = append(locations, apistructs.PipelineTypeCICD.String()+string(os.PathSeparator))
```
Comment (sfwn) [inline modules/pipeline/services/extmarketsvc/common.go:41]: why default is cicd at pipeline side?

**Review by sfwn — CHANGES_REQUESTED** [review]
where is your recursive-query logic of location?

**Comment by kakj-go** [comment]
> where is your recursive-query logic of location?

For the time being, I am still considering whether it is necessary to recursively query based on the value of location, or whether to let the upper layer pass it by itself

File: modules/pipeline/services/extmarketsvc/common.go [hunk 5]
```diff
@@ -25,3 +31,26 @@ func MakeActionTypeVersion(action *pipelineyml.Action) string {
...
 	return r
 }
+
+func MakeActionLocationsBySource(source apistructs.PipelineSource) []string {
+	var locations []string
+	switch source {
+	case apistructs.PipelineSourceCDPDev, apistructs.PipelineSourceCDPTest, apistructs.PipelineSourceCDPStaging, apistructs.PipelineSourceCDPProd, apistructs.PipelineSourceBigData:
+		locations = append(locations, apistructs.PipelineTypeFDP.String()+string(os.PathSeparator))
+	case apistructs.PipelineSourceDice, apistructs.PipelineSourceProject, apistructs.PipelineSourceProjectLocal, apistructs.PipelineSourceOps, apistructs.PipelineSourceQA:
+		locations = append(locations, apistructs.PipelineTypeCICD.String()+string(os.PathSeparator))
```
Comment (sfwn) [inline modules/pipeline/services/extmarketsvc/common.go:41]: do not use `os.PathSeparator`. It's OS depends. In windows, is `\\`.

File: modules/pipeline/services/extmarketsvc/list.go [hunk 6]
```diff
@@ -163,7 +167,7 @@ func (s *ExtMarketSvc) continuousRefreshAction() {
 
 // each search item: ActionType 或 ActionType@version
 // output: map[item]Image
-func (s *ExtMarketSvc) SearchActions(items []string, ops ...OpOption) (map[string]*diceyml.Job, map[string]*apistructs.ActionSpec, error) {
+func (s *ExtMarketSvc) SearchActions(items []string, location []string, ops ...OpOption) (map[string]*diceyml.Job, map[string]*apistructs.ActionSpec, error) {
```
Comment (sfwn) [inline modules/pipeline/services/extmarketsvc/list.go:170]: locations

**Review by sfwn — CHANGES_REQUESTED** [review]

**Comment by sfwn** [comment]
> 
Pass the locations explicitly is better.
And we can provider some strategy suck as `recursive find by '/'` in pipeline side later.

**Review by sfwn — APPROVED** [review]

**Pull request merged**